File: tools/authority-campaign-publisher/src/island_mountain_publisher/linkedin.py
"""Small LinkedIn REST client used by the NOOA publisher."""

from __future__ import annotations

import json
import time
import urllib.parse
import urllib.request
from collections.abc import Callable
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from .authority import MutationKind, MutationReceipt
from .models import ManifestItem


class LinkedInError(RuntimeError):
    """LinkedIn rejected or did not confirm a publication step."""


class LinkedInClient:
    """Upload one card, publish its summary, then add the article comment."""

    def __init__(
        self,
        repository_root: Path,
        access_token: str,
        actor_urn: str,
        api_version: str,
        *,
        sleep: Callable[[float], None] = time.sleep,
    ) -> None:
        if not actor_urn.startswith("urn:li:person:"):
            raise LinkedInError("LINKEDIN_ACTOR_URN must identify Basho's person account")
        self.repository_root = repository_root
        self.access_token = access_token
        self.actor_urn = actor_urn
        self.api_version = api_version
        self.sleep = sleep

    def _headers(self) -> dict[str, str]:
        return {
            "Authorization": f"Bearer {self.access_token}",
            "LinkedIn-Version": self.api_version,
            "X-Restli-Protocol-Version": "2.0.0",
            "Content-Type": "application/json",
        }

    def _request(
        self,
        method: str,
        url: str,
        payload: dict[str, Any] | bytes | None = None,
        *,
        headers: dict[str, str] | None = None,
    ) -> tuple[dict[str, Any], dict[str, str]]:
        body = (
            json.dumps(payload).encode("utf-8")
            if isinstance(payload, dict)
            else payload
        )
        request = urllib.request.Request(
            url,
            data=body,
            method=method,
            headers=headers or self._headers(),
        )
        try:
            with urllib.request.urlopen(request, timeout=45) as response:
                raw = response.read()
                parsed = json.loads(raw) if raw else {}
                return parsed, {key.lower(): value for key, value in response.headers.items()}
        except Exception as exc:
            raise LinkedInError(f"LinkedIn {method} did not complete: {url}") from exc

    def upload_image(self, item: ManifestItem) -> MutationReceipt:
        if item.campaign.card_path is None:
            raise LinkedInError(f"{item.campaign.campaign_id} has no card")
        initialized, _ = self._request(
            "POST",
            "https://api.linkedin.com/rest/images?action=initializeUpload",
            {"initializeUploadRequest": {"owner": self.actor_urn}},
        )
        value = initialized.get("value", {})
        upload_url = value.get("uploadUrl")
        image_urn = value.get("image")
        if not upload_url or not image_urn:
            raise LinkedInError("LinkedIn did not return an image upload target")
        card = (self.repository_root / item.campaign.card_path).read_bytes()
        self._request(
            "PUT",
            str(upload_url),
            card,
            headers={
                "Authorization": f"Bearer {self.access_token}",
                "Content-Type": "application/octet-stream",
            },
        )
        encoded = urllib.parse.quote(str(image_urn), safe="")
        for _attempt in range(12):
            image, _ = self._request(
                "GET",
                f"https://api.linkedin.com/rest/images/{encoded}",
            )
            if image.get("status") == "AVAILABLE":
                return MutationReceipt(
                    campaign_id=item.campaign.campaign_id,
                    kind=MutationKind.LINKEDIN_IMAGE,
                    remote_id=str(image_urn),
                    occurred_at=datetime.now(UTC),
                )
            self.sleep(5)
        raise LinkedInError("LinkedIn image did not become available")

    def create_post(self, item: ManifestItem, image_urn: str) -> MutationReceipt:
        commentary = "\n\n".join(
            (
                item.campaign.linkedin_summary,
                " ".join(item.campaign.linkedin_hashtags),
            )
        )
        _body, headers = self._request(
            "POST",
            "https://api.linkedin.com/rest/posts",
            {
                "author": self.actor_urn,
                "commentary": commentary,
                "visibility": "PUBLIC",
                "distribution": {
                    "feedDistribution": "MAIN_FEED",
                    "targetEntities": [],
                    "thirdPartyDistributionChannels": [],
                },
                "content": {
                    "media": {
                        "id": image_urn,
                        "title": item.campaign.title,
                        "altText": item.alt_text,
                    }
                },
                "lifecycleState": "PUBLISHED",
                "isReshareDisabledByAuthor": False,
            },
        )
        post_urn = headers.get("x-restli-id")
        if not post_urn:
            raise LinkedInError("LinkedIn created no identifiable post")
        return MutationReceipt(
            campaign_id=item.campaign.campaign_id,
            kind=MutationKind.LINKEDIN_POST,
            remote_id=post_urn,
            occurred_at=datetime.now(UTC),
        )

    def create_comment(self, item: ManifestItem, post_urn: str) -> MutationReceipt:
        encoded = urllib.parse.quote(post_urn, safe="")
        _body, headers = self._request(
            "POST",
            f"https://api.linkedin.com/rest/socialActions/{encoded}/comments",
            {
                "actor": self.actor_urn,
                "message": {"text": item.linkedin_first_comment},
            },
        )
        comment_id = headers.get("x-restli-id")
        if not comment_id:
            raise LinkedInError("LinkedIn created no identifiable first comment")
        return MutationReceipt(
            campaign_id=item.campaign.campaign_id,
            kind=MutationKind.LINKEDIN_COMMENT,
            remote_id=comment_id,
            occurred_at=datetime.now(UTC),
        )

File: worker/src/publisher/linkedin.ts
/**
 * LinkedIn REST client — TypeScript port of island_mountain_publisher.linkedin.
 * Upload the campaign card, publish the summary post, then add the article first comment.
 * The comment uses the /v2 socialActions API (the versioned /rest one is partner-gated for
 * member tokens — proven 2026-08-10). Requires the three LINKEDIN_* secrets.
 */
import type { Env } from '../types'
import type { CampaignDay } from './content'
import { readFileBytes } from './github'

const REST = 'https://api.linkedin.com/rest'
const V2 = 'https://api.linkedin.com/v2'

function headers(env: Env): Record<string, string> {
  return {
    Authorization: `Bearer ${env.LINKEDIN_ACCESS_TOKEN}`,
    'LinkedIn-Version': env.LINKEDIN_VERSION as string,
    'X-Restli-Protocol-Version': '2.0.0',
    'Content-Type': 'application/json'
  }
}

async function fail(prefix: string, res: Response): Promise<never> {
  throw new Error(`${prefix} ${res.status}: ${(await res.text()).slice(0, 200)}`)
}

export async function uploadImage(env: Env, day: CampaignDay): Promise<string> {
  const actor = env.LINKEDIN_ACTOR_URN as string
  if (!day.linkedin.card_path) throw new Error(`${day.campaign_id}: no card`)
  const initRes = await fetch(`${REST}/images?action=initializeUpload`, {
    method: 'POST',
    headers: headers(env),
    body: JSON.stringify({ initializeUploadRequest: { owner: actor } })
  })
  if (!initRes.ok) await fail('LinkedIn initializeUpload', initRes)
  const init = (await initRes.json()) as { value?: { uploadUrl?: string; image?: string } }
  const uploadUrl = init.value?.uploadUrl
  const imageUrn = init.value?.image
  if (!uploadUrl || !imageUrn) throw new Error('LinkedIn returned no image upload target')
  const bytes = await readFileBytes(env, day.linkedin.card_path, 'main')
  const put = await fetch(uploadUrl, {
    method: 'PUT',
    headers: {
      Authorization: `Bearer ${env.LINKEDIN_ACCESS_TOKEN}`,
      'Content-Type': 'application/octet-stream'
    },
    body: bytes
  })
  if (!put.ok) await fail('LinkedIn image PUT', put)
  const encoded = encodeURIComponent(imageUrn)
  for (let attempt = 0; attempt < 12; attempt += 1) {
    const check = await fetch(`${REST}/images/${encoded}`, { headers: headers(env) })
    if (check.ok) {
      const img = (await check.json()) as { status?: string }
      if (img.status === 'AVAILABLE') return imageUrn
    }
    await new Promise((resolve) => setTimeout(resolve, 5000))
  }
  throw new Error('LinkedIn image did not become AVAILABLE')
}

export async function createPost(env: Env, day: CampaignDay, imageUrn: string): Promise<string> {
  const res = await fetch(`${REST}/posts`, {
    method: 'POST',
    headers: headers(env),
    body: JSON.stringify({
      author: env.LINKEDIN_ACTOR_URN,
      commentary: day.linkedin.commentary,
      visibility: 'PUBLIC',
      distribution: {
        feedDistribution: 'MAIN_FEED',
        targetEntities: [],
        thirdPartyDistributionChannels: []
      },
      content: {
        media: { id: imageUrn, title: day.linkedin.title, altText: day.linkedin.alt_text }
      },
      lifecycleState: 'PUBLISHED',
      isReshareDisabledByAuthor: false
    })
  })
  if (!res.ok) await fail('LinkedIn createPost', res)
  const urn = res.headers.get('x-restli-id')
  if (!urn) throw new Error('LinkedIn created no identifiable post')
  return urn
}

export async function createComment(env: Env, day: CampaignDay, postUrn: string): Promise<string> {
  const encoded = encodeURIComponent(postUrn)
  const res = await fetch(`${V2}/socialActions/${encoded}/comments`, {
    method: 'POST',
    headers: headers(env),
    body: JSON.stringify({
      actor: env.LINKEDIN_ACTOR_URN,
      message: { text: day.linkedin.first_comment }
    })
  })
  if (!res.ok) await fail('LinkedIn createComment', res)
  const id = res.headers.get('x-restli-id')
  if (!id) throw new Error('LinkedIn created no identifiable comment')
  return id
}
